solr/core/src/java/org/apache/solr/core/TransientSolrCoreCacheDefault.java

```diff
@@ -70,10 +80,15 @@ public TransientSolrCoreCacheDefault(CoreContainer coreContainer) {
     Caffeine<String, SolrCore> transientCoresCacheBuilder =
         Caffeine.newBuilder()
             .initialCapacity(initialCapacity)
-            // Use the current thread to queue evicted cores for closing. This ensures the
-            // cache max size is respected (with a different thread the max size would be
-            // respected asynchronously only eventually).
-            .executor(Runnable::run)
+            // Do NOT use the current thread to queue evicted cores for closing. Although this
+            // ensures the cache max size is respected only eventually, it should be a very
+            // short period of time until the cache maintenance task kicks in.
+            // The onEvict method needs the writeLock from SolrCores to operate safely.
+            // However, the current thread most probably has acquired a read-lock already
+            // somewhere up the call stack and would deadlock.
+            // Note that Caffeine cache has an internal maintenance task rescheduling logic which
+            // explicitly checks on the common pool.
+            .executor(ForkJoinPool.commonPool())
```

Comment:
I am still reluctant to introduce an eviction approver as I believe it is an easy but wrong answer. It is perfectly pragmatic for a private api where the caveats are known, but for a library it is a poor api with footguns that will have a large blast radius.

I would use a secondary cache with weak referenced values for automatic clean up. That looks similar to the overflow cache here, except that by using reference counting it is discarded by the GC. This way there are no memory leaks and the same instance is retrievable if any other thread holds a strong reference to it to avoid duplications due to races. If simpler then the second cache can be inclusive since it is based on what is GC reachable rather than a size/time bound, so juggling the exclusivity might be useless logic that does not save any memory. By using Map computations the interactions between caches can be atomic, e.g. to load from the weak cache if absent from the primary one.

A caveat is that a weak value means that a listener on that cache would have a null value (as GC'd), so it may not be immediately compatible with `SolrCores.queueCoreToClose(core)`. It may be that refactoring to coordinate the close based on reachability could solve that and remove the complexity that `SolrCores` must go through to safely shutdown an instance.

Reply:
Here SolrCore instances are not usual object hold in memory, they can be still in-use, they are ref-counted, there can be pending operations on them. We have to close them carefully, not GCed as weak references.
It becomes clearer to me that this transient core cache cannot have a strict limit, because it must manage the cores until they are not used anymore and unloaded. It's more a target size, that the cache should aim to keep when there is no heavy activity on the cores.
If we go with an overflow structure, it is probably not a cache but an unlimited size map, with its maintenance thread to monitor the cores to close when they are not used anymore.
Or we keep a single cache from which we do not remove cores if they are still in-use (so not Caffeine). But we would still need a maintenance thread to reduce the cache size down to the target size when it is possible.
I'll take some time to reflect on this and probably create a Jira issue on this subject.

Reply:
It’s too bad the api is not like a managed resource, like a stream or jdbc connection. Then you could return a proxy that pins it when in use, unpins when released. The proxy doing the callbacks like a jdbc pool or txn manager keeps it hidden from the application logic from handling explicitly.

Reply:
Reviewing the API to be able to pin is indeed another possible approach.
When we pin a Caffeine cache entry (with weight 0), what operations are involved internally? Does it reorder some kind of structure (priority queue) for eviction? Or is it very cheap?

Reply:
Caffeine only uses amortized O(1) data structures so it is relatively cheap. The entry is on an lru doubly linked list that is managed in a non-blocking fashion. Currently the zero weights are left on the lru and skipped over during eviction, but they could later be moved on/off if that traversal became an issue. The modifying of an entry’s weight is done only by write operations, e.g. on a put, so more expensive that a read but still [very fast](https://github.com/ben-manes/caffeine/wiki/Benchmarks#write-100-1) (45M/s).